BH-1075 Allow creation of taxonomies in the developer app

nickcernis opened the pull request:
- Adds a new taxonomies view at `/wp-admin/?page=atlas-content-modeler&view=taxonomies`.
- Adds a “View Taxonomies” button on the main model listings page (for now; this may move later).

<img width="1470" alt="Screenshot 2021-06-30 at 11 43 32" src="https://user-images.githubusercontent.com/647669/123939499-78ceb080-d998-11eb-809e-f672ed98d13b.png">

<img width="1452" alt="Screenshot 2021-06-30 at 11 43 21" src="https://user-images.githubusercontent.com/647669/123939512-7c623780-d998-11eb-98eb-f1a500bcc97c.png">

### To test
1. Visit the developer app at `/wp-admin/admin.php?page=atlas-content-modeler`.
2. Click “View Taxonomies”.
3. Try adding taxonomies. You should see them in the list to the right of the taxonomies form.
4. Run `wp option get atlas_content_modeler_taxonomies` or view that row in the WP options table to see the result of your submission as data.
 
### Notes
- We plan to improve the listing of taxonomies in [BH-1078](https://wpengine.atlassian.net/browse/BH-1078) (“List Taxonomies”). I added the listing area to the right of the form as a placeholder for now just to wire up the data.
- Perhaps we could re-use the existing form in [BH-1076](https://wpengine.atlassian.net/browse/BH-1076) (“Edit Taxonomy”)? We could hide the list view in edit mode and present only the form with submit/cancel buttons.

nickcernis committed:
chore: add button for Taxonomies view

This may be relocated later or replaced with a
different nav system. We just need an easy way
of accessing the taxonomies page for now.

nickcernis committed:
chore: add taxonomies reducer

The main models reducer is already fairly complex
but well scoped to just model and field data.
Keeping the taxonomies reducer separate should
help us to manage and maintain model and taxonomy
state independently.

nickcernis committed:
feat: add taxonomies view

To display ACM taxonomies and add new ones.

nickcernis committed:
chore: improve prop name

nickcernis committed:
chore: add create taxonomies form

In progress. This does not submit to a REST
endpoint yet.

nickcernis committed:
style: adjust heading and checklist styling

nickcernis committed:
chore: make string translatable

nickcernis committed:
feat: register create taxonomy REST endpoint

The REST API file is now ripe for refactoring;
we could pull model, field and taxonomy endpoints
into different files in a future PR.

nickcernis committed:
fix: 'type' → 'types' to match taxonomy spec

nickcernis committed:
chore: create addTaxonomy reducer action

So that the app immediately reflects taxonomy
state after successful taxonomy creation.

nickcernis committed:
chore: handle new taxonomy form submission

Success results in:
 - Resetting the form to its empty state.
 - Showing a toast.
 - Scrolling to the top.
 - Updating the Taxonomies list.

nickcernis committed:
fix: reset the API ID field on submit success

nickcernis committed:
refactor: allow taxonomy update via PUT request

Not strictly needed for the “add taxonomy” PR,
but will make future work on editing taxonomies
easier and required minimal changes.

nickcernis committed:
test: verify the taxonomy REST endpoint

nickcernis committed:
fix: amend check for at least one model

Sites with only one ACM model will present a
single checkbox in the list of models.

This is submitted as a string, not an array, so
we need to cast strings to arrays for consistent
storage.

Unchecked models are sent either as boolean
false (only one model) or an empty array (multiple
models), so this also amends the check for those
cases.

nickcernis committed:
chore: add taxonomy page acceptance test helper

nickcernis committed:
fix: relink fields on successful form submission

When someone edits the taxonomy API Identifier
manually, this breaks the “link” between the
Singular name field and the API Identifier field
(so that edits to the singular name no longer
update the API ID field, because this would
destroy any manual changes they made).

When someone edits the API ID field manually
(breaking the link) and then saves their new
taxonomy successfully, this link between the
fields needs to be reset so the fields are linked
again for the next taxonomy to be entered.

This ensures that the next taxonomy someone adds
will generate an API ID from the Singular field
again.

We did not have this issue with other forms,
because the form for models/fields does not remain
in view after submission.

nickcernis committed:
test: add e2e tests for the Taxonomy view

mindctrl commented on includes/rest-api/rest-api-endpoint-registration.php:
```diff
@@ -680,3 +724,60 @@ function content_model_multi_option_exists( array $names, string $current_choice
+		);
+	}
+
+	$taxonomies = get_option( 'atlas_content_modeler_taxonomies', array() );
+
+	if ( ! $is_update && array_key_exists( $params['slug'], $taxonomies ) ) {
+		return new WP_Error(
+			'atlas_content_modeler_taxonomy_exists',
+			esc_html__( 'A taxonomy with this API Identifier already exists.', 'atlas-content-modeler' ),
+			[ 'status' => 400 ]
+		);
+	}
```
Nice. We should probably also check for conflicts with existing taxonomies registered via other means (WP core, other plugins, custom code, etc.)

nickcernis replied:
Good idea. I think this should cover it: https://github.com/wpengine/atlas-content-modeler/pull/183/commits/3fadd02e8e940c2fa050af854c8dda2ee68e1f47.

mindctrl commented on includes/rest-api/rest-api-endpoint-registration.php:
```diff
@@ -680,3 +724,60 @@ function content_model_multi_option_exists( array $names, string $current_choice
+ *
+ * @param array $params Parameters passed from the taxonomy form.
+ * @param bool  $is_update True if `$params` came from a PUT request.
+ * @return object|WP_Error
```
```suggestion
 * @return array|WP_Error
```

It looks like `$taxonomy` is an array

nickcernis replied:
Good catch, thanks — not sure why PhpStorm didn't infer that. 🤷

nickcernis committed:
chore: update phpdoc return type

Co-authored-by: John Parris <john.parris@wpengine.com>

nickcernis committed:
fix: check for WP core taxonomy collisions

Addresses https://github.com/wpengine/atlas-content-modeler/pull/183#discussion_r661537397.

Props @mindctrl.

mindctrl commented on includes/settings/js/src/components/Taxonomies.jsx:
```diff
@@ -0,0 +1,505 @@
+import Icon from "../../../../components/icons";
+import { useApiIdGenerator } from "./fields/useApiIdGenerator";
+import { showSuccess } from "../toasts";
+import { toValidApiId } from "./fields/toValidApiId";
```
```suggestion

```
Looks like this is not used now.

mindctrl commented on includes/settings/js/src/components/Taxonomies.jsx:
```diff
@@ -0,0 +1,505 @@
+									className="w-100"
+									ref={register({
+										required: true,
+										maxLength: 20,
```
```suggestion
										maxLength: 32,
```

`register_taxonomy` supports 32 chars here and we went with that in the schema def here: https://github.com/wpengine/atlas-content-modeler/blob/ab37aad2c043603e2d6b86319fa29b567750a163/docs/adr/0002-taxonomy-definition-schema.md

mindctrl commented on includes/settings/js/src/components/Taxonomies.jsx:
```diff
@@ -0,0 +1,505 @@
+												<Icon type="error" />
+												<span role="alert">
+													{__(
+														"Exceeds max length of 20.",
```
```suggestion
														"Exceeds max length of 32",
```

to match above suggestion

mindctrl reviewed (changes requested):
Overall working well! Just a few comments.

nickcernis committed:
chore: update max length description to 32

Co-authored-by: John Parris <john.parris@wpengine.com>

nickcernis committed:
chore: update max length to 32

Co-authored-by: John Parris <john.parris@wpengine.com>

nickcernis committed:
refactor: remove unused import

Co-authored-by: John Parris <john.parris@wpengine.com>

nickcernis committed:
refactor: adjust taxonomy dupe detection logic

$wp_taxonomies will eventually include ACM
taxonomies once the taxonomy registration PR is
merged.

Previous behavior would have then seen all PUT
requests fail because $wp_taxonomies would have
included the ACM taxonomies.

This commit adjusts “slug exists” logic to:

- Block creation or update of a taxonomy if one
  with the same slug exists that was not created
  by ACM.
- Block creation of a taxonomy if one with the
  same slug exists that was created by ACM.
- Allow updates to a taxonomy, but only if it was
  created by ACM.

nickcernis committed:
Merge branch 'bh-1075-taxonomy-definition' of https://github.com/wpengine/atlas-content-modeler into bh-1075-taxonomy-definition

apmatthews commented on includes/settings/js/src/components/Taxonomies.jsx:
```diff
@@ -0,0 +1,504 @@
+					});
+					window.scrollTo(0, 0);
+					reset();
+					setApiIdGeneratorInput(""); // Resets the API ID field.
```
We also need to run `setSingularCount(0)` and `setPluralCount(0)` as part of the reset.

nickcernis replied:
Thank you, well spotted! Fixed in 76b2e04.

apmatthews reviewed (commented):
Just one minor comment, but works well otherwise!

nickcernis committed:
fix: reset field character counters on success

Props @apmatthews.

Addresses https://github.com/wpengine/atlas-content-modeler/pull/183#discussion_r661827869.
